docs/blog/2021-06-21-zephyr-2-5.md

```diff
@@ -0,0 +1,59 @@
+---
+title: "Zephyr 2.5 Update"
+author: Pete Johanson
+author_title: Project Creator
+author_url: https://gitlab.com/petejohanson
+author_image_url: https://www.gravatar.com/avatar/2001ceff7e9dc753cf96fcb2e6f41110
+tags: [firmware, zephyr, core]
+---
+
+Happy to announce that we have completed the work to upgrade ZMK to [Zephyr 2.5](https://docs.zephyrproject.org/2.5.0/releases/release-notes-2.5.html)!
+
+A big part of this work was some _major_ refactors and improvements by [innovaker] to our [zmk-docker](https://github.com/zmkfirmware/zmk-docker/) Docker image and GH Actions automation. With the updated changes there, we can now:
+
+- GH and layer caching for improved build times.
+- Automated tests run against newly built images to verify them.
+- PRs to the repo now properly build and run the tests as well.
+- Build images for multiple target architectures, e.g. `zmk-build-riscv64`, all in parallel.
+- Create nightly builds to be sure we're pulling in the latest OS/package updates, to ensure we keep our images up to date, address any reported vulnerabilities, etc.
+- Allow us to upgrade to newer Zephyr SDK and Zephyr versions, more quickly.
+- Publish candidate tags, and then promote tested ones to final tags like `zmk-dev-arm:2.5`.
+
+## Getting The Changes
+
+Use the following steps to update to the latest tooling in order to properly use the new ZMK changes:
+
+### Github Actions
+
+Existing Github Actions will pull down Zephyr 2.5 automatically, and should work, fine as is. However, to be safe, you should edit the `.github/workflows/build.yml` to change `zmkfirmware/zmk-build-arm:2.4` to `zmkfirmware/zmk-build-arm:2.5`.
```

Comment:
AFAICT, it should be a drop in replacement to switch tot hew newer docker image, so I will add a note about that.

Reply:
If so, I'd suggest recommending that they replace their `build.yml` with the new syntax anyway because it's a lot simpler to understand.  i.e. point them at:
https://github.com/zmkfirmware/zmk-config-template/blob/main/.github/workflows/build.yml
or
https://github.com/zmkfirmware/zmk-config-split-template/blob/main/.github/workflows/build.yml
(which both need updating come to think of it)

Reply:
That said, I guess the placeholders in the templates aren't immediately obvious.

Reply:
This seems reasonably recent if you need a test case ...
https://github.com/krikun98/zmk-config/blob/master/.github/workflows/build.yml

Reply:
> If so, I'd suggest recommending that they replace their `build.yml` with the new syntax anyway because it's a lot simpler to understand. i.e. point them at:
> https://github.com/zmkfirmware/zmk-config-template/blob/main/.github/workflows/build.yml
> or
> https://github.com/zmkfirmware/zmk-config-split-template/blob/main/.github/workflows/build.yml
> (which both need updating come to think of it)

I don't think we should be suggesting this, *lots* of our users are new to `git` and GitHub, so asking them to refactor portions of some obscure YAML file is *not* productive, IMHO. I'd rather we give them a simple "replace X with Y" that anyone can follow easily.

Reply:
Fair enough.  Happy to kick this bucket down the road some more.

Reply:
> AFAICT, it should be a drop in replacement to switch tot hew newer docker image, so I will add a note about that.

As in ... https://github.com/innovaker/zmk-config/actions/runs/958253475/workflow ... ?

Reply:
On the bright side, `zephyr-west-action-arm` does build the 2.5 branch without complaining.

Reply:
The recent improvements to the instructions cover anyone using `zmk-build-arm:2.4` but I still think those using `zephyr-west-action-arm` are going to be perplexed when they can't find `zmk-build-arm` in their `build.yml`.  As such, I'd imagine ...
- some users just ignore it and leave their `build.yml` as is.
- some users will attempt to substitute `zephyr-west-action-arm` with `zmk-build-arm:2.5` which won't work (see above).
- some users will fall into #help.

2.5 isn't the problem, but it compounds the underlying issue that the zmk-config setup/templating system wasn't designed to facilitate upgrades.  And I suspect 2.6 even more so because it has dependency changes.

Short of a new setup approach, I suggest nipping this in the bud with a migration guide, migration script or a note to remake their `zmk-config` repository.  At minimum a note in the blog highlighting this possible scenario to the longtime users.

Reply:
Yeah, I think a note to highlight this, and suggest re-generating their config repo and dropping back in their keymaps, etc if they want to upgrade. I don't think investing time in migration scripts is worthwhile at this point.

Reply:
WFM.